### PR TITLE
Add support and doc for AD4858

### PR DESCRIPTION
**Pull request by mphalke**
Added initial precision toolbox support for AD4858 ADC

### CODE REVIEW

File: +adi/+AD4858/Rx.m [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    methods (Hidden, Access = protected)
+
+        function setupInit(obj)
+            % Write all attributes to device once connected through set
+            % methods
+            % Do writes directly to hardware without using set methods.
+            % This is required since Simulink support doesn't support
+            % modification to nontunable variables at SetupImpl
+
+            % obj.setDeviceAttributeRAW('sampling_frequency',num2str(obj.SampleRate));
```
Comment (tfcollins) [inline +adi/+AD4858/Rx.m:88]: Why is this commented?

File: +adi/+AD4858/Rx.m [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        function obj = Rx(varargin)
+            obj = obj@matlabshared.libiio.base(varargin{:});
+            obj.enableExplicitPolling = false;
+            obj.EnabledChannels = 1;
+            obj.BufferTypeConversionEnable = true;
+            obj.uri = 'ip:analog.local';
+        end
+
+        function flush(obj)
+            flushBuffers(obj);
```
Comment (tfcollins) [inline +adi/+AD4858/Rx.m:65]: Needs a docstring

File: +adi/+AD4858/Rx.m [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    %   rx = adi.AD4858.Rx('uri','192.168.2.1');
+    %
+    %   <a href="https://www.analog.com/media/en/technical-documentation/data-sheets/ad4858.pdf">AD4858 Datasheet</a>
+
+    properties (Nontunable)
+        % SampleRate Sample Rate
+        %   Baseband sampling rate in Hz, specified as a scalar
+        %   in samples per second.
+        SampleRate = '500000'
+        SamplesPerFrame = 1024
```
Comment (tfcollins) [inline +adi/+AD4858/Rx.m:17]: Needs docstring

**Review by tfcollins — CHANGES_REQUESTED** [review]
Why does the AD4858 class have a custom stepImpl?

File: +adi/+AD4858/Rx.m (hunk 1 above)
Comment (mphalke) [reply +adi/+AD4858/Rx.m:88]: will uncomment this

**Comment by mphalke** [comment]
> stepImpl

I took other existing projects as a baseline for this (e.g. ad7380, ad7768) and all of them have this custom stepImpl. So I used the same for ad4858

**Comment by mphalke** [comment]
V1.0: Addressed review comments

1. Added docstrings as per review comments
2. Uncommented sampling_frequency set function call in setupInit

**Comment by tfcollins** [comment]
> > stepImpl
> 
> I took other existing projects as a baseline for this (e.g. ad7380, ad7768) and all of them have this custom stepImpl. So I used the same for ad4858

Ok that probably got missed since the commit was so large (176 files). Anyway, can you just use the built in stepImpl ?

**Comment by ribdp** [comment]
> > > stepImpl
> > 
> > 
> > I took other existing projects as a baseline for this (e.g. ad7380, ad7768) and all of them have this custom stepImpl. So I used the same for ad4858
> 
> Ok that probably got missed since the commit was so large (176 files). Anyway, can you just use the built in stepImpl ?

In the precision toolbox, another 'FrameCount' parameter was added which the user can use to specify upfront the number of buffers of data requested. The reasoning was that all those buffers can be captured in one go, and the formatting/type conversion is done afterwards, so as not to add any additional delays between buffers. 

I wasn't too sure if this should be upstreamed to ToolboxCommon, so I put that in the individual drivers for the time being. But that's not ideal either, so I'm taking out the duplicated `stepImpl` definitions from the drivers, and putting it (along with some other stuff) in a common base class, in another PR.

**Comment by mphalke** [comment]
@tfcollins Any further inputs on this PR?

**Comment by tfcollins** [comment]
> > > > stepImpl
> > > 
> > > 
> > > I took other existing projects as a baseline for this (e.g. ad7380, ad7768) and all of them have this custom stepImpl. So I used the same for ad4858
> > 
> > 
> > Ok that probably got missed since the commit was so large (176 files). Anyway, can you just use the built in stepImpl ?
> 
> In the precision toolbox, another 'FrameCount' parameter was added which the user can use to specify upfront the number of buffers of data requested. The reasoning was that all those buffers can be captured in one go, and the formatting/type conversion is done afterwards, so as not to add any additional delays between buffers.
> 
> I wasn't too sure if this should be upstreamed to ToolboxCommon, so I put that in the individual drivers for the time being. But that's not ideal either, so I'm taking out the duplicated `stepImpl` definitions from the drivers, and putting it (along with some other stuff) in a common base class, in another PR.

So will help somewhat within a frame group but makes it worse across groups. This also doesn't guarantee anything since you can still generate gaps in data by just increasing the sample rate. If speed is a huge concern we can update MATLAB to use the new iio_channel_read which doesn't require multiple data copies. This is what pyadi-iio does now.

Is there a frame size requirement you are trying to get to so that data is contiguous?

-Travis

**Comment by ribdp** [comment]
> > > > > stepImpl
> > > > 
> > > > 
> > > > I took other existing projects as a baseline for this (e.g. ad7380, ad7768) and all of them have this custom stepImpl. So I used the same for ad4858
> > > 
> > > 
> > > Ok that probably got missed since the commit was so large (176 files). Anyway, can you just use the built in stepImpl ?
> > 
> > 
> > In the precision toolbox, another 'FrameCount' parameter was added which the user can use to specify upfront the number of buffers of data requested. The reasoning was that all those buffers can be captured in one go, and the formatting/type conversion is done afterwards, so as not to add any additional delays between buffers.
> > I wasn't too sure if this should be upstreamed to ToolboxCommon, so I put that in the individual drivers for the time being. But that's not ideal either, so I'm taking out the duplicated `stepImpl` definitions from the drivers, and putting it (along with some other stuff) in a common base class, in another PR.
> 
> So will help somewhat within a frame group but makes it worse across groups. This also doesn't guarantee anything since you can still generate gaps in data by just increasing the sample rate. If speed is a huge concern we can update MATLAB to use the new iio_channel_read which doesn't require multiple data copies. This is what pyadi-iio does now.
> 
> Is there a frame size requirement you are trying to get to so that data is contiguous?
> 
> -Travis

That this would worsen the delay between frame groups, does make sense. While there's no specific frame size I was going after, the intention was just to speed things up within the frame group, and move another step towards 'continuous' capture across, potentially, more than one MAX_SIZE buffers. 

If we can get a speed upgrade by switching to iio_channel_read, that should be great.

**Comment by mphalke** [comment]
V2.0: Removed the custom implementation of stepImpl
-Removed cutom stepImpl  implementation from AD4858 Rx base class

**Comment by mphalke** [comment]
@tfcollins This PR is ready to review now. Please review.

File: +adi/+AD4858/Rx.m (hunk 2 above)
Comment (tfcollins) [reply +adi/+AD4858/Rx.m:70]: Still missing docstring

**Review by tfcollins — CHANGES_REQUESTED** [review]
flush missing docstring

File: +adi/+AD4858/Rx.m (hunk 2 above)
Comment (mphalke) [reply +adi/+AD4858/Rx.m:70]: When I see the diff, I see that docstring has already been added. Could you double check?

**Comment by mphalke** [comment]
> flush missing docstring

When I see the diff, I see that docstring has already been added. Could you double check?

**Comment by tfcollins** [comment]
Still on there. Just to verify, by docstring in MATLAB this is text that is within the function at the top and appears with the help command

**Comment by mphalke** [comment]
> Still not there. Just to verify, by docstring in MATLAB this is text that is within the function at the top and appears with the help command

got it now, have corrected this

**Comment by ribdp** [comment]
To generate documentation for this part, could you run the `gen_sysobj_doc.m` file present in CI\gen_doc\docs\ and make one final push? This would update sysobjs.json with the doc strings from your driver - which in turn will be used in the automatic documentation build

**Comment by mphalke** [comment]
V3.0: AD4858: Generated documentation
-Auto generated documentation .json file by running gen_sysobj_doc.m file

**Review by ribdp — APPROVED** [review]

**Pull request merged**